pkg/mcs/resource_manager/server/token_bukets.go

```diff
@@ -38,107 +41,259 @@ type GroupTokenBucket struct {
+}
+
 // NewGroupTokenBucket returns a new GroupTokenBucket
-func NewGroupTokenBucket(tokenBucket *rmpb.TokenBucket) GroupTokenBucket {
+func NewGroupTokenBucket(tokenBucket *rmpb.TokenBucket) *GroupTokenBucket {
 	if tokenBucket == nil || tokenBucket.Settings == nil {
-		return GroupTokenBucket{}
+		return &GroupTokenBucket{}
 	}
-	return GroupTokenBucket{
-		Settings: tokenBucket.Settings,
+	return &GroupTokenBucket{
+		Settings: tokenBucket.GetSettings(),
 		GroupTokenBucketState: GroupTokenBucketState{
-			Tokens: tokenBucket.Tokens,
+			Tokens:     tokenBucket.GetTokens(),
+			tokenSlots: make(map[uint64]*TokenSlot),
 		},
 	}
 }
 
 // GetTokenBucket returns the grpc protoc struct of GroupTokenBucket.
-func (t *GroupTokenBucket) GetTokenBucket() *rmpb.TokenBucket {
-	if t.Settings == nil {
+func (gtb *GroupTokenBucket) GetTokenBucket() *rmpb.TokenBucket {
+	if gtb.Settings == nil {
 		return nil
 	}
 	return &rmpb.TokenBucket{
-		Settings: t.Settings,
-		Tokens:   t.Tokens,
+		Settings: gtb.Settings,
+		Tokens:   gtb.Tokens,
 	}
 }
 
 // patch patches the token bucket settings.
-func (t *GroupTokenBucket) patch(tb *rmpb.TokenBucket) {
+func (gtb *GroupTokenBucket) patch(tb *rmpb.TokenBucket) {
 	if tb == nil {
 		return
 	}
 	if setting := proto.Clone(tb.GetSettings()).(*rmpb.TokenLimitSettings); setting != nil {
-		t.Settings = setting
-		t.settingChanged = true
+		gtb.Settings = setting
+		gtb.settingChanged = true
 	}
 
-	// the settings in token is delta of the last update and now.
-	t.Tokens += tb.GetTokens()
+	// The settings in token is delta of the last update and now.
+	gtb.Tokens += tb.GetTokens()
 }
 
 // init initializes the group token bucket.
-func (t *GroupTokenBucket) init(now time.Time) {
-	if t.Settings.FillRate == 0 {
-		t.Settings.FillRate = defaultRefillRate
+func (gtb *GroupTokenBucket) init(now time.Time, clientID uint64) {
+	if gtb.Settings.FillRate == 0 {
+		gtb.Settings.FillRate = defaultRefillRate
 	}
-	if t.Tokens < defaultInitialTokens {
-		t.Tokens = defaultInitialTokens
+	if gtb.Tokens < defaultInitialTokens {
+		gtb.Tokens = defaultInitialTokens
 	}
-	t.LastUpdate = &now
-	t.Initialized = true
+	// init slot
+	gtb.tokenSlots[clientID] = &TokenSlot{
+		settings:          gtb.Settings,
+		tokenCapacity:     gtb.Tokens,
+		lastTokenCapacity: gtb.Tokens,
+		lastUpdate:        &now,
+	}
+	gtb.LastUpdate = &now
+	gtb.Initialized = true
 }
 
-// request requests tokens from the group token bucket.
-func (t *GroupTokenBucket) request(now time.Time, neededTokens float64, targetPeriodMs uint64) (*rmpb.TokenBucket, int64) {
-	if !t.Initialized {
-		t.init(now)
-	} else {
-		delta := now.Sub(*t.LastUpdate)
-		if delta > 0 {
-			t.Tokens += float64(t.Settings.FillRate) * delta.Seconds()
-			t.LastUpdate = &now
-		}
+// updateTokens updates the tokens and settings.
+func (gtb *GroupTokenBucket) updateTokens(now time.Time, burstLimit int64, clientUniqueID uint64, consumptionToken float64) {
+	var elapseTokens float64
```

Comment:
I think there can be an optimization here. Some clients may consume RU at a low speed, resulting in that token exceeds burst and but the entire Resource group cannot reach the upper limit of RU consumption.